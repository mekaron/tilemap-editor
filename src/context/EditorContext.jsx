import React, { createContext, useReducer, useContext } from 'react';

export const TOOLS = {
  BRUSH: 0,
  ERASE: 1,
  PAN: 2,
  PICK: 3,
  RAND: 4,
  FILL: 5,
};

export const initialState = {
  prevActiveTool: TOOLS.BRUSH,
  activeTool: TOOLS.BRUSH,
  activeMap: '',
  displaySymbols: false,
  showGrid: false,
  selection: [],
  currentLayer: 0,
  isMouseDown: false,
  maps: {},
  tileSets: {},
  activeLoader: 'default',
};

export const editorReducer = (state, action) => {
  switch (action.type) {
    case 'SET_TOOL':
      return {
        ...state,
        prevActiveTool: state.activeTool,
        activeTool: action.tool,
      };
    case 'ADD_MAP':
      return {
        ...state,
        maps: { ...state.maps, [action.map.name]: action.map },
        activeMap: action.map.name,
      };
    case 'UPDATE_LAYER': {
      const { mapName, layerIndex, layerData } = action;
      const map = state.maps[mapName];
      if (!map) return state;
      const layers = map.layers.map((l, i) =>
        i === layerIndex ? { ...l, ...layerData } : l
      );
      return {
        ...state,
        maps: { ...state.maps, [mapName]: { ...map, layers } },
      };
    }
    case 'SET_ACTIVE_MAP':
      return { ...state, activeMap: action.name };
    case 'TOGGLE_GRID':
      return { ...state, showGrid: !state.showGrid };
    default:
      return state;
  }
};

const EditorContext = createContext();

export const EditorProvider = ({ children }) => {
  const [state, dispatch] = useReducer(editorReducer, initialState);
  return (
    <EditorContext.Provider value={{ state, dispatch, TOOLS }}>
      {children}
    </EditorContext.Provider>
  );
};

export const useEditor = () => useContext(EditorContext);

export default EditorContext;
