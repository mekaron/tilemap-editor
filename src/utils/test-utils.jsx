import React from 'react';
import { render } from '@testing-library/react';
import EditorProvider from '../context/EditorProvider';

const AllTheProviders = ({ children }) => {
  return <EditorProvider>{children}</EditorProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
