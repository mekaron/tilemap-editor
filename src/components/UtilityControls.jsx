import React, { useContext } from 'react';
import EditorContext from '../context/EditorContext';

export default function UtilityControls() {
  const { editorState, setEditorState } = useContext(EditorContext);
  const { zoom } = editorState;

  const handleZoomIn = () => {
    setEditorState({ ...editorState, zoom: zoom * 2 });
  };

  const handleZoomOut = () => {
    setEditorState({ ...editorState, zoom: zoom / 2 });
  };

  return (
    <div className="tool_wrapper">
      <label id="undoBtn" title="Undo">↩️️</label>
      <label id="redoBtn" title="Redo">🔁️</label>
      <label id="zoomIn" title="Zoom in" onClick={handleZoomIn}>🔎️+</label>
      <label id="zoomOut" title="Zoom out" onClick={handleZoomOut}>🔎️-</label>
      <label id="zoomLabel">{zoom}x️</label>
    </div>
  );
}
