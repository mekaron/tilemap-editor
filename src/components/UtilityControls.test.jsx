import React from 'react';
import { render, screen, fireEvent } from '../utils/test-utils';
import UtilityControls from './UtilityControls';
import EditorContext from '../context/EditorContext';

const mockEditorState = {
  zoom: 1,
};

const setEditorState = jest.fn();

const renderComponent = () =>
  render(
    <EditorContext.Provider value={{ editorState: mockEditorState, setEditorState }}>
      <UtilityControls />
    </EditorContext.Provider>
  );

test('UtilityControls renders correctly', () => {
  const { container } = renderComponent();
  expect(screen.getByTitle('Zoom in')).toBeTruthy();
  expect(container.querySelector('.tool_wrapper')).toMatchSnapshot();
});

test('UtilityControls changes zoom on click', () => {
  renderComponent();
  const zoomInButton = screen.getByTitle('Zoom in');
  fireEvent.click(zoomInButton);
  expect(setEditorState).toHaveBeenCalledWith({
    ...mockEditorState,
    zoom: 2,
  });

  const zoomOutButton = screen.getByTitle('Zoom out');
  fireEvent.click(zoomOutButton);
  expect(setEditorState).toHaveBeenCalledWith({
    ...mockEditorState,
    zoom: 0.5,
  });
});
