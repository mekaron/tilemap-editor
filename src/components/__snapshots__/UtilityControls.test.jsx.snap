// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UtilityControls renders correctly 1`] = `
<div
  class="tool_wrapper"
>
  <label
    id="undoBtn"
    title="Undo"
  >
    ↩️️
  </label>
  <label
    id="redoBtn"
    title="Redo"
  >
    🔁️
  </label>
  <label
    id="zoomIn"
    title="Zoom in"
  >
    🔎️+
  </label>
  <label
    id="zoomOut"
    title="Zoom out"
  >
    🔎️-
  </label>
  <label
    id="zoomLabel"
  >
    1
    x️
  </label>
</div>
`;
